src/librustc/hir/lowering.rs

```diff
@@ -2992,10 +2992,11 @@ impl<'a> LoweringContext<'a> {
         hir::PatKind::Wild
     }
 
-    fn lower_range_end(&mut self, e: &RangeEnd) -> hir::RangeEnd {
+    fn lower_range_end(&mut self, e: &RangeEnd, has_end: bool) -> hir::RangeEnd {
         match *e {
-            RangeEnd::Included(_) => hir::RangeEnd::Included,
-            RangeEnd::Excluded => hir::RangeEnd::Excluded,
+            RangeEnd::Excluded if has_end => hir::RangeEnd::Excluded,
+            // No end; so `X..` behaves like `RangeFrom`.
+            RangeEnd::Excluded | RangeEnd::Included(_) => hir::RangeEnd::Included,
```

Comment:
So the cases that are unrespresentable in syntax (but still well-defined semantically in HIR itself) are:
- `Range(None, None, _)` -- syntax taken by rest patterns `..`, semantically equivalent to `_`
- `Range(Some(_), None, Excluded)` -- doesn't have a syntax either

I feel that's not so bad and the upside is that a) this fits the existing structure better (I started changing HIR and the fallout seemed like it would have deep consequences), so it's less of an upheaval, b) the logic in HAIR, typeck, and beyond becomes more uniform.

As for AST & the parser, I think `Bound` doesn't fit with the fact that we need e.g. `RangeEnd::Included(RangeSyntax::DotDotDot)` and that there's a `Spanned<_>` in there.

All in all, I would prefer to keep things as-is, since this keeps the PR mostly a shallow surface-centric change. If we want to refactor HIR + HAIR we could always do so later.

Reply:
>  If we want to refactor HIR + HAIR we could always do so later.

agreed